types/test.ts

```diff
@@ -5,24 +5,24 @@ import '@testing-library/cypress/add-commands'
 configure({testIdAttribute: 'data-myown-testid'})
 
 // findBy*
-cy.findByPlaceholderText('foo') // $ExpectType Chainable<JQuery<HTMLElement>>
```

Comment:
Why should this type signature change? You'd have to update all the return types of every command in the `index.d.ts` file. All `JQuery` commands return `JQuery<HTMLElement>` by default. It would make more sense to update the signature of all queries to be generic with a default:

```ts
findAllByText<T = HTMLElement>(
  id: Matcher,
  options?: MatcherOptions
): Chainable<JQuery<T>>
```

This way you could override the interface if you need to:
```ts
cy.findAllByText<SVGTextElement>('some text')
```

I don't consider this to be the safest option, which is why I made `cy.filter` accept tags as both a runtime and type guard: https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts#L965

Unfortunately that `filter` is extending the key only from `HTMLElementTagNameMap` which excludes SVG elements. That type interface could be expanded to include any element map so that `cy.filter` can give the correct interface.

What do you think?

Reply:
Sorry this is going a bit over my head as I have limited experience with Cypress. The intention was to port https://github.com/testing-library/react-testing-library/pull/833 so that SVG an other elements would be allowed in Cypress tests as well. Is this not the right way to do that? I assumed `JQuery<Element>` would also work on SVG elements.

Reply:
Cypress is fundamentally a different testing environment than React Testing Library:

- There is no `get*` or `query*` queries
  - All Cypress commands are always asynchronous, so `get*` queries don't make sense in Cypress
  - Cypress seamlessly handles non-existence, so `query*` queries don't make sense. `cy.findByRole('button').should('not.exist')` works without issue
- Cypress Testing Library doesn't have a `render`. Cypress works against a running web page. That can be a fully-working application, Storybook via [cypress-storybook](https://github.com/NicholasBoll/cypress-storybook), or unit-testing via [cypress-react-unit-test](https://github.com/bahmutov/cypress-react-unit-test) (which has it's own `mount`).
- Cypress's assertions use [chai-jquery](https://docs.cypress.io/guides/references/assertions.html#Chai-jQuery) which has a lot of powerful assertion types that aren't element-specific and the element interface doesn't actually matter.
- Cypress chains wrap jQuery which has it's own API separate from DOM APIs. https://github.com/testing-library/react-testing-library/issues/830 doesn't apply. I think the following is okay though: https://github.com/testing-library/cypress-testing-library/pull/164/files#diff-093ad82a25aee498b11febf1cdcb6546e4d223ffcb49ed69cc275ac27ce0ccceR13, although that isn't the idiomatic way to scope queries in Cypress:
```ts
// idiomatic React Testing Library
const container = screen.getByRole('menu')
expect(screen.queryByRole('menuitem', { container, name: 'My Menu Item' })).toBeNull()

// non-idiomatic Cypress Testing Library using idiomatic React Testing Library patterns


// idiomatic Cypress Testing Library
cy.findByRole('menu')
  .findByRole('menuitem', { name: 'My Menu Item'  })
  .should('not.exist')
```

Cypress or jQuery doesn't actually care what the underlying elements are and Cypress Testing Library doesn't have any mechanism of rendering anything, so there is no type definition issue there.

I think the types can be expanded though, but that doesn't have anything to do with https://github.com/testing-library/react-testing-library/pull/833:

- `container` can be expanded to `Element | JQuery<Element>` like you have. Although not the preferred method for scoping queries in Cypress Testing Library, there's no reason to limit the interface to `HTMLElement`
- Make all `find*` queries generic so in the rare instances where the interface needs to be specific to make Typescript happy it is possible without hacks:

```ts
findAllByText<T = HTMLElement>( // default to `HTMLElement` for backwards compatibility and a good default
  id: Matcher,
  options?: MatcherOptions
): Chainable<JQuery<T>>
```

This way it is possible to do something like:
```ts
cy.findByRole<HTMLButtonElement>('button')
  .then(el => { // `el` is now `jQuery<HTMLButtonElement>`
    const form = el[0].form // this property exists according to Typescript. Unfortunately it can return `null`
  })
```

This isn't the best example as the following is a bit easier to read:
```ts
cy.findByRole('button')
  .parents('form') // better - will either return a `form` element, or test will fail stating `parents` could not find selector `form`
```

Cypress traversal commands match jQuery traversal methods.

But the generic interface will allow type overriding in the rare instance it is needed.

Reply:
I think it's simplest to revert this file then. A generic isn't necessary here as it only sets the return type, which can just be casted by the user explicitly.